cosmos/providers/dbt/core/operators/virtualenv.py

```diff
@@ -0,0 +1,146 @@
+from airflow.utils.python_virtualenv import prepare_virtualenv
+
+from cosmos.providers.dbt.core.operators.local import (
+    DbtDocsLocalOperator,
+    DbtLocalBaseOperator,
+    DbtLSLocalOperator,
+    DbtRunLocalOperator,
+    DbtRunOperationLocalOperator,
+    DbtSeedLocalOperator,
+    DbtSnapshotLocalOperator,
+    DbtTestLocalOperator,
+)
+
+logger = logging.getLogger(__name__)
+
+
+PY_INTERPRETER = "python3"
+
+
+class DbtVirtualenvBaseOperator(DbtLocalBaseOperator):
+    """
+    Executes a dbt core cli command within a Python Virtual Environment, that is created before running the DBT command
+    and deleted at the end of the operator execution.
+
+    :param py_requirements: If defined, creates a virtual environment with the specified dependencies. Example:
+           ["dbt-postgres==1.5.0"]
+    :param py_system_site_packages: Whether or not all the Python packages from the Airflow instance will be accessible
+           within the virtual environment (if py_requirements argument is specified).
+           Avoid using unless the DBT job requires it.
+    """
+
+    def __init__(
+        self,
+        py_requirements: list[str] | None = None,
+        py_system_site_packages: bool = False,
+        **kwargs,
+    ) -> None:
+        self.py_requirements = py_requirements or []
+        self.py_system_site_packages = py_system_site_packages
+        super().__init__(**kwargs)
+        self._venv_tmp_dir = ""
+
+    @cached_property
+    def venv_dbt_path(
+        self,
+    ) -> str:
+        """
+        Path to the DBT binary within a Python virtualenv.
+
+        The first time this property is called, it creates a virtualenv and installs the dependencies based on the
+        self.py_requirements and self.py_system_site_packages. This value is cached for future calls.
+        """
+        # We are reusing the virtualenv directory for all subprocess calls within this task/operator.
+        # For this reason, we are not using contexts at this point.
+        # The deletion of this directory is done explicitly at the end of the `execute` method.
+        self._venv_tmp_dir = TemporaryDirectory(prefix="cosmos-venv")
+        py_interpreter = prepare_virtualenv(
+            venv_directory=self._venv_tmp_dir.name,
+            python_bin=PY_INTERPRETER,
+            system_site_packages=self.py_system_site_packages,
+            requirements=self.py_requirements,
+        )
+        dbt_binary = Path(py_interpreter).parent / "dbt"
+        cmd_output = self.subprocess_hook.run_command(
+            [
+                py_interpreter,
+                "-c",
+                "from importlib.metadata import version; print(version('dbt-core'))",
+            ]
+        )
+        dbt_version = cmd_output.output
+        self.log.info("Using DBT version %s available at %s", dbt_version, dbt_binary)
+        return str(dbt_binary)
+
+    def run_subprocess(self, command, *args, **kwargs):
+        if self.py_requirements:
```

Comment:
Should there logging for the command being executed here like there is in the DbtLocalBaseOperator?

Reply:
Since the log is currently in `run_command` and the `VirtualenvOperator` is not overriding it, it is printed in both cases. I noticed some duplication in logging the command since the Subprocess hook also logs the command from `run_command` (but without `the `cwd`). I'll be improving this in a separate PR!